### ISSUE TITLE
Finish/document OpFromGraph

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#1603]
We have a meta op called OpFromGraph() defined in theano/compile/builders.py.

This is thing that are relatively fast to do and are needed to make it usable:
- [ ] Document it so that user know how to user it.
- [x] test if it work with shared variable as input and if not make it work. Add test for that.
- [ ] Move it to the GPU. We can do it quickly for the old back-end, move all float32 inputs to the GPU. Otherwise, we need to compile the inner function, see which inputs get moved to the GPU, then create a new OpFromGraph with the corresponding input to the GPU. https://github.com/Theano/Theano/pull/2982

This is think that should be done to make it complete, but don't prevent people to use it:
- [x] Move the grad work in the grad() method. This should remove the grad_depth parameter
- [x] Make the grad use only 1  op for all outputs, otherwise, this would duplicate computation.
- [x] Opt to unfold it
- [x] implement infer_shape gh-3117
- [x] implement Rop
- [x] Allow the grad and Rop to be overrided.
- [ ] Opt to remove unused inputs/outputs (even more useful when better grad)
- [ ] implement c_code when all ops have c code.
- [ ] Implement __hash__ and __eq__
- [ ] Make the grad return DisconnectedType
- [ ] updates? givens?
- [ ] test pickle of this op
- [ ]  test random numbers with this op.

This is more an ideas that should be discussed before implementation:
- [ ] Allow the inner graph to work inplace on the inputs?
- [ ] implement optimization to remove that op to the global graph when it is small? This allow better optimization

I put it tentatively in 0.6.1.

**Comment by nouiz** [comment]
I started a branch some time ago that started that, it isn't finished. I do not plan to work on it in the short term:

https://github.com/nouiz/Theano/tree/op_from_graph

**Comment by nouiz** [comment]
For the next part, can you finish my PR gh-2982? I don't have time for this and I think it is one of the main missing feature.

**Comment by ChienliMa** [comment]
OK, I try to understand your codes first.

**Comment by ChienliMa** [comment]
I just realize that I have some problem to finish this task. 

The worst one is that I cannot reach any PC with Nvidia GPU. I don't think I can run the test without one Nvidia card.  I've apply for a GPU instance on AWS, and I will inform you if I can work on it.

Do you think I can finish this task without a PC with Nvidia GPU? 
If not, I think it's better to switch to another task.

**Comment by Sentient07** [comment]
@nouiz , from the commit message, I guess shared variable as input is done and the test has been added. Can that be marked as done ?

**Comment by Sentient07** [comment]
@nouiz in this PR, https://github.com/Theano/Theano/pull/4039 , i am using props_dict() for GPUOPs. Since i am using the __props__ you have asked me to remove the existing __hash__ and __eq__ methods. I guess it also applies here and we don't need them anymore ?

**Comment by nouiz** [comment]
Don't touch OpFromGraph. It is complicated and not used currently. So it is better to not to change it until we continue it more. I think OpFromGraph will need special \__hash__ and \__eq__method.

SharedVariable are now supported. I marked it as done. thanks.

**Comment by Sentient07** [comment]
@nouiz Also, on moving it to the GPU, we need to follow the latter method, right ?

**Comment by nouiz** [comment]
I'm not sure what you mean by the latter method. But the PR that allow this op to work on GPU isn't merged. So on the issue of moving op to the GPU faster, it don't need to touch OpFromGraph. I think it is better that you don't touch it. I need much work to be useful and is not the current direction. As it don't work now on GPU, you don't need to do anything about it.

**Comment by Sentient07** [comment]
In the sub-task 
> Move it correctly to the GPU. 

there are two approaches you have talked about there. Could i proceed with the second approach ?

> Otherwise, we need to compile the inner function, see which inputs get moved to the GPU, then create a new OpFromGraph with the corresponding input to the GPU.

**Comment by nouiz** [comment]
Exact. We don't move this op to the GPU now, so you don't need to change it.

**Comment by khaotik** [comment]
Just to note this list needs updating after #5255.

**Comment by nouiz** [comment]
I updated the description.
